aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,871 @@
+#   form they were parsed from.  These structures are immutable
+#   and you're expected to populate information once during
+#   construction.
+#
+# - Strict mypy typechecking.  You can typecheck this file using
+#   `mypy -config mypy-strict.ini aten/src/ATen/gen_cpu.py` and
+#   this will enforce that everything is annotated.
+#
+# Some non-goals:
+#
+# - Change native_functions.yaml format.  One step at a time!
+#
+# The general structure:
+#
+# - We define a lot of dataclasses to represent all of the various
+#   semantic entities in native_functions.yaml (schema! types!)
+#   These classes come with parsing and pretty-printing functionality.
+#
+# - We parse native_functions.yaml into our dataclasses
+#
+# - We do code generation on it (under construction!)
+#
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                           DATA MODEL
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# The basic input to the code generation is native_functions.yaml.
+# The name "native", BTW, comes from the distinction between native
+# functions and legacy TH functions.  The legacy TH functions are gone,
+# but the "native" descriptor has stuck.
+#
+# NativeFunction models a single entry in native_functions.yaml.  Its
+# fields roughly correspond to what you would see in the YAML itself,
+# but after canonicalization and parsing has occurred.
+#
+# You can see some of the overall design patterns for how we setup
+# dataclasses in this class, but we will defer a complete discussion
+# of this at FunctionSchema.
+@dataclass(frozen=True)
+class NativeFunction:
+    # The function schema of the operator in question.  This schema
+    # has been parsed; see FunctionSchema for more about its structure.
+    # (This type is quoted as we are forward referencing a type
+    # defined later in the file.  I opted for this ordering of the
+    # classes for expository clarity.)
+    func: 'FunctionSchema'
+
+    # Corresponds to the 'use_c10_dispatcher' field.  Historically,
+    # this field could take several possible strings, but right
+    # now you can have it in any color you like, as long as it's 'full'
+    use_c10_dispatcher_full: bool
+
+    # If no variants are specified in native_functions.yaml, this is
+    # assumed to be {'function'}.
+    variants: Set['Variant']
+
+    # Whether or not we should skip generating registrations for
+    # this kernel.  This is a bit of a double-edged sword, as manual
+    # registrations don't participate in codegen-based selective build!
+    manual_kernel_registration: bool
+
+    # Distinguish between a missing dispatch dict (historically, this
+    # means to register a catch-all kernel) and a present but empty
+    # dispatch dict (this means register nothing; arguably, this should
+    # subsume manual_kernel_registration).
+    #
+    # TODO: str key could be replaced with more explicit enum
```

Comment:
This is an interesting one - it would be great to drive both codegen and c++ defs off the same source of truth, especially if we want to use the more complicated relationships we're contemplating, like aliases. (And unless we do unify the source of truth, a more semantic datatype will add change-tracking overhead in any case.)